src/gil.rs

```diff
@@ -153,54 +153,56 @@ impl Drop for GILGuard {
 
 /// Thread-safe storage for objects which were dropped while the GIL was not held.
 struct ReleasePool {
-    pointers_to_drop: Mutex<*mut Vec<NonNull<ffi::PyObject>>>,
-    pointers_being_dropped: UnsafeCell<*mut Vec<NonNull<ffi::PyObject>>>,
+    locked: AtomicBool,
+    pointers_to_drop: UnsafeCell<Vec<NonNull<ffi::PyObject>>>,
+}
+
+struct Lock<'a> {
+    lock: &'a AtomicBool,
+}
+
+impl<'a> Lock<'a> {
+    fn new(lock: &'a AtomicBool) -> Self {
+        while lock.compare_and_swap(false, true, Ordering::Acquire) {
+            spin_loop_hint();
+        }
+        Self { lock }
+    }
+}
+
+impl<'a> Drop for Lock<'a> {
+    fn drop(&mut self) {
+        self.lock.store(false, Ordering::Release);
+    }
 }
 
 impl ReleasePool {
     const fn new() -> Self {
         Self {
-            pointers_to_drop: parking_lot::const_mutex(std::ptr::null_mut()),
-            pointers_being_dropped: UnsafeCell::new(std::ptr::null_mut()),
+            locked: AtomicBool::new(false),
+            pointers_to_drop: UnsafeCell::new(Vec::new()),
         }
     }
 
     fn register_pointer(&self, obj: NonNull<ffi::PyObject>) {
-        let mut storage = self.pointers_to_drop.lock();
-        if storage.is_null() {
-            *storage = Box::into_raw(Box::new(Vec::with_capacity(256)))
-        }
+        let _lock = Lock::new(&self.locked);
+        let v = self.pointers_to_drop.get();
         unsafe {
-            (**storage).push(obj);
+            if (*v).capacity() == 0 {
+                (*v).reserve(256);
```

Comment:
Recently I noticed this kind of reserving doesn't offer so much speed up.
Could you please remove this if sentences?

Reply:
Sure. Done.